hail/src/main/scala/is/hail/expr/ir/functions/LocusFunctions.scala

```diff
@@ -342,132 +246,205 @@ object LocusFunctions extends RegistryFunctions {
         PCanonicalStruct("locus" -> PCanonicalLocus(lTyp.rg, true), "alleles" -> PCanonicalArray(PCanonicalString(true), true))
       }
     }) {
-      case (r, rt: PStruct, (strT, variantoff: Code[Long])) =>
-        val plocus = rt.types(0).asInstanceOf[PLocus]
-        val svar = asm4s.coerce[String](wrapArg(r, strT)(variantoff))
+      case (r, cb, rt: PCanonicalStruct, variantStr) =>
+
+        val svar = variantStr.asString.loadString()
+        val plocus = rt.types(0).asInstanceOf[PCanonicalLocus]
         val variant = Code
           .invokeScalaObject2[String, ReferenceGenome, (Locus, IndexedSeq[String])](
-          VariantMethods.getClass, "parse", svar, rgCode(r.mb, plocus.rg))
-        emitVariant(r, variant, rt)
+            VariantMethods.getClass, "parse", svar, rgCode(r.mb, plocus.rg))
+        emitVariant(cb, r.region, variant, rt)
     }
 
-    registerEmitCode2("LocusInterval", TString, TBoolean, tinterval("T"), {
+    registerIEmitCode2("LocusInterval", TString, TBoolean, tinterval("T"), {
       (returnType: Type, _: PType, _: PType) => {
         val lPTyp = returnType.asInstanceOf[TInterval].pointType.asInstanceOf[TLocus]
         PCanonicalInterval(PCanonicalLocus(lPTyp.asInstanceOf[TLocus].rg))
       }
-    }) {
-      case (r: EmitRegion, rt: PInterval, ioff: EmitCode, invalidMissing: EmitCode) =>
-        val plocus = rt.pointType.asInstanceOf[PLocus]
-        val sinterval = asm4s.coerce[String](wrapArg(r, ioff.pt)(ioff.value[Long]))
-        val intervalLocal = r.mb.newLocal[Interval](name="intervalObject")
-        val interval = Code.invokeScalaObject3[String, ReferenceGenome, Boolean, Interval](
-          locusClass, "parseInterval", sinterval, rgCode(r.mb, plocus.rg), invalidMissing.value[Boolean])
-
-        EmitCode(
-          Code(ioff.setup, invalidMissing.setup),
-          ioff.m || invalidMissing.m || Code(intervalLocal := interval, intervalLocal.isNull),
-          PCode(rt, emitInterval(r, intervalLocal, rt))
-        )
+    }) { case (cb: EmitCodeBuilder, r: Value[Region], rt: PCanonicalInterval, locusStrEC: EmitCode, invalidMissingEC: EmitCode) =>
+      val plocus = rt.pointType.asInstanceOf[PLocus]
+
+
+      locusStrEC.toI(cb).flatMap(cb) { locusStr =>
+        invalidMissingEC.toI(cb).flatMap(cb) { invalidMissing =>
+
+          val Lmissing = CodeLabel()
+          val Ldefined = CodeLabel()
+
+          val interval = cb.newLocal[Interval]("locus_interval_interval",
+            Code.invokeScalaObject3[String, ReferenceGenome, Boolean, Interval](
+              locusClass, "parseInterval",
+              locusStr.asString.loadString(),
+              rgCode(cb.emb, plocus.rg),
+              invalidMissing.asBoolean.boolCode(cb)))
+
+          cb.ifx(interval.isNull, cb.goto(Lmissing))
+
+          val intervalCode = emitLocusInterval(cb, r, interval, rt)
+          cb.goto(Ldefined)
+          IEmitCode(Lmissing, Ldefined, intervalCode)
```

Comment:
A slightly simpler way to do this:
```
IEmitCode(cb, interval.isNull, interval)
  .map(cb)(emitLocusInterval(cb, r, _, rt))
```

There's several other places you could use the same pattern, but up to you which style you find clearer.

Reply:
I want to push toward removing the generic type on IEmitCode in favor of SCode as soon as we can. I guess we could do something like:
```
IEmitCode(cb, interval.isNull, {
... stuff here ... })
```

Which would do the same. I'm probably not going to go back and change it though, if it's OK with you

Reply:
That's fine, I'll approve.